pkg/controller/core.oam.dev/v1alpha2/application/revision.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2021 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package application
+
+import (
+	"context"
+
+	"github.com/google/go-cmp/cmp"
+
+	"github.com/pkg/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha2"
+	"github.com/oam-dev/kubevela/pkg/appfile"
+	"github.com/oam-dev/kubevela/pkg/controller/utils"
+)
+
+// GenerateRevision will generate revision for an Application when created/updated
+func (r *Reconciler) GenerateRevision(ctx context.Context, appfile *appfile.Appfile, application *v1alpha2.Application, ac *v1alpha2.ApplicationConfiguration, comps []*v1alpha2.Component) (bool, *v1alpha2.AppRevisionHash, error) {
+	appHash, err := utils.ComputeSpecHash(application.Spec)
+	if err != nil {
+		return false, nil, err
+	}
+	var componentDefinitionHash, traitDefinitionHash []v1alpha2.NameAndHashValue
+
+	copiedApp := application.DeepCopy()
+	// clean up the status of application going to be recorded in revision
+	copiedApp.Status = v1alpha2.AppStatus{}
+	appRev := &v1alpha2.ApplicationRevision{
+		Spec: v1alpha2.ApplicationRevisionSpec{
+			Application:              copiedApp,
+			ComponentDefinitions:     nil, // TODO: add componentDefinition when completed.
+			Components:               comps,
+			ApplicationConfiguration: ac,
+		},
+	}
+	appRev.Name = appfile.RevisionName
+	appRev.Namespace = application.Namespace
+
+	for _, w := range appfile.Workloads {
+		if w == nil {
+			continue
+		}
+		if !DupWorkloadDef(appRev, w.FullTemplate.WorkloadDefinition) {
```

Comment:
I don't think we need this but if we have to, why not make them a map[definiiton]string instead of an array? We are clearly using it as a map.